trie/trie.go

```diff
@@ -306,36 +342,68 @@ func (t *Trie) insert(n node, prefix, key []byte, value node) (bool, node, error
 	}
 	switch n := n.(type) {
 	case *shortNode:
+		if t.isStorageTrie && t.currentEpoch >= 2 {
+			if expired, err := t.nodeExpired(n, prefix); expired {
+				return false, n, err
+			}
+		}
 		matchlen := prefixLen(key, n.Key)
 		// If the whole key matches, keep this short node as is
 		// and only update the value.
 		if matchlen == len(n.Key) {
-			dirty, nn, err := t.insert(n.Val, append(prefix, key[:matchlen]...), key[matchlen:], value)
+			n.setEpoch(t.currentEpoch)
+			dirty, nn, err := t.insert(n.Val, append(prefix, key[:matchlen]...), key[matchlen:], value, n.epoch)
 			if !dirty || err != nil {
 				return false, n, err
 			}
-			return true, &shortNode{Key: n.Key, Val: nn, flags: t.newFlag()}, nil
+			return true, &shortNode{Key: n.Key, Val: nn, flags: t.newFlag(), epoch: t.currentEpoch}, nil
 		}
 		// Otherwise branch out at the index where they differ.
 		branch := &fullNode{flags: t.newFlag()}
 		var err error
-		_, branch.Children[n.Key[matchlen]], err = t.insert(nil, append(prefix, n.Key[:matchlen+1]...), n.Key[matchlen+1:], n.Val)
+		_, branch.Children[n.Key[matchlen]], err = t.insert(nil, append(prefix, n.Key[:matchlen+1]...), n.Key[matchlen+1:], n.Val, t.currentEpoch)
 		if err != nil {
 			return false, nil, err
 		}
-		_, branch.Children[key[matchlen]], err = t.insert(nil, append(prefix, key[:matchlen+1]...), key[matchlen+1:], value)
+		if t.isStorageTrie {
+			branch.setEpoch(t.currentEpoch)
+			branch.UpdateChildEpoch(int(n.Key[matchlen]), t.currentEpoch)
+		}
+		_, branch.Children[key[matchlen]], err = t.insert(nil, append(prefix, key[:matchlen+1]...), key[matchlen+1:], value, t.currentEpoch)
 		if err != nil {
 			return false, nil, err
 		}
+		if t.isStorageTrie {
+			branch.setEpoch(t.currentEpoch)
+			branch.UpdateChildEpoch(int(key[matchlen]), t.currentEpoch)
+		}
 		// Replace this shortNode with the branch if it occurs at index 0.
 		if matchlen == 0 {
 			return true, branch, nil
 		}
 		// Otherwise, replace it with a short node leading up to the branch.
-		return true, &shortNode{Key: key[:matchlen], Val: branch, flags: t.newFlag()}, nil
+		return true, &shortNode{Key: key[:matchlen], Val: branch, flags: t.newFlag(), epoch: t.currentEpoch}, nil
 
 	case *fullNode:
-		dirty, nn, err := t.insert(n.Children[key[0]], append(prefix, key[0]), key[1:], value)
+		if t.isStorageTrie {
+			if t.currentEpoch >= 2 {
+				// this full node is expired, return err
+				if expired, err := t.nodeExpired(n, prefix); expired {
+					return false, n, err
+				}
+			}
+			// else, set its epoch to current epoch.
+			n.setEpoch(t.currentEpoch)
+			if t.currentEpoch >= 2 {
```

Comment:
this check may unnecessary, epoch is uint16, and current epoch must great then trie epoch, or check in childExpired function?